### PR TITLE
Use FormChoiceFormatter on more providers

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | Uses automatic form choice formatter on more things in the backoffice. It prevents duplicate values from being shown, automatically sorts values.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Tests should be green.
| UI Tests          | https://github.com/florine2623/testing_pr/actions/runs/11813874605 :x:
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

**Review by FabienPapet — DISMISSED** [review]
Nice improvement. Perhaps this will point out that FormChoiceProviders are useless :p

**Comment by Hlavtox** [comment]
@FabienPapet https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Form/ChoiceProvider/ShippingInclusionChoiceProvider.php

🤣

**Comment by Hlavtox** [comment]
@FabienPapet Restarting UI tests for a 5th time now, let's see if it finally finishes. :-)

**Review by ShaiMagal — DISMISSED** [review]

**Review by FabienPapet — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @Hlavtox ,

I've relaunched the tests multiple times. They're still red.

ping @PrestaShop/qa-automation is theres something to change from @Hlavtox here ?

Thanks!

**Comment by kpodemski** [comment]
@Hlavtox 

They are related to selecting options, so I suggest checking if you didn't break something.

**Comment by Hlavtox** [comment]
@florine2623 @kpodemski @PrestaShop/qa-automation 

There was one flipped key that I fixed, but I checked the other things and I can't find anything wrong.
- Order messages work properly.
- Payment vs group config works.
- Tax settings work.

Let's see what tests say. :-)

**Review by kpodemski — DISMISSED** [review]

**Comment by Hlavtox** [comment]
New test run https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11878767287

**Comment by Hlavtox** [comment]
@PrestaShop/qa-automation Guys, could you please check the two failures from this run - https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11878767287 and adapt the tests to it? I did not find anything wrong in my testing. Thank you!

**Comment by florine2623** [comment]
Hello @Hlavtox ,

I have relaunched the auto tests with the latest updates done by automation : https://github.com/florine2623/testing_pr/actions/runs/11969884252

It is still not green 😓 

As seen with @PrestaShop/qa-automation , you need to update your PR to make the auto tests work.

Could you check ?
Thanks!

**Comment by kpodemski** [comment]
@florine2623 @PrestaShop/qa-automation 

I'm not sure about this test:

```
  1) BO - International - Taxes : Edit Tax options with all EcoTax values
       Edit tax options
         should edit Tax Option,
      	Enable Tax:true,
      	Display tax in the shopping cart: 'false',
      	Based on: 'Delivery address',
      	Use ecotax: 'true',
      	Ecotax: 'EU VAT For Virtual Products':
     locator.selectOption: Timeout 30000ms exceeded.
```
     
When I go to International -> Taxes -> I edit the group `EU VAT For Virtual Products`, when I click [here](https://www.dropbox.com/s/rvu625jogwgln37/Zrzut%20ekranu%202024-11-22%20o%2014.26.21.png?dl=0), I have a failed ajax request due to the error:

>  "detail": "Call to protected method AdminTaxRulesGroupControllerCore::displayAjaxupdateTaxRule() from scope PrestaShopBundle\\Controller\\Admin\\LegacyController",

@florine2623 do you know about this problem? The exact same error exists on `9.0.x` without @Hlavtox modifications.

And for this test `BO - Payment - Preferences : Configure group restrictions`, I also couldn't reproduce it.

I successfully disabled `ps_wirepayment` for the `Customer` group, the module disappeared in front office.

**Comment by Hlavtox** [comment]
@kpodemski Exactly, I can't find any issue with this PR.

**Comment by florine2623** [comment]
@kpodemski ,

Without this PR, I also have the same issue as you :

https://github.com/user-attachments/assets/16032fec-9940-42be-a644-8451134bea30

**Comment by Hlavtox** [comment]
@PrestaShop/qa-functional @PrestaShop/qa-functional 

Guys, so, what's the result, can I merge it or not? The PR works as confirmed by @kpodemski.

**Comment by Progi1984** [comment]
I'm here for UI Tests.

Scenario `tests/UI/campaigns/functional/BO/11_international/03_taxes/01_taxes/04_taxOptionsForm.ts` : 

The option value for Tax Options > Ecotax has changed : 
* Before : `EU VAT For Virtual Products `
* After : `EU VAT For Virtual Products (5)`

image

Scenario `tests/UI/campaigns/functional/BO/10_payment/02_preferences/02_groupRestrictions.ts` : 

The order of group has changed : 
* Before : <br/> image
* After : <br/> image

**Comment by Hlavtox** [comment]
OK, thanks @Progi1984, we moved! :-)

> The option value for Tax Options > Ecotax has changed :

There were items with duplicate name and ID. I handled this case.

vat

> The order of group has changed :

I preserved the order of groups as they are returned from the provider.

Should be good now. :-)

**Comment by kpodemski** [comment]
thank you @Progi1984 ❤️

**Comment by Hlavtox** [comment]
@florine2623 @Progi1984 Yaaaay, tests are green! :-)

https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11878767287

**Review by kpodemski — APPROVED** [review]

**Comment by Progi1984** [comment]
@Hlavtox Thanks to UI Tests :) They make their job.

File: src/Core/Form/FormChoiceFormatter.php [hunk 1]
```diff
@@ -50,13 +50,25 @@ class FormChoiceFormatter
      */
     public static function formatFormChoices(array $rawChoices, string $idKey, string $nameKey, bool $sortByName = true): array
     {
+        // Remove items with duplicate name and ID
```
Comment (FabienPapet) [inline src/Core/Form/FormChoiceFormatter.php:53]: Is there a test that covers this use case ? 

Can be added later, but this will make the PR perfect :p

Comment (Hlavtox) [reply src/Core/Form/FormChoiceFormatter.php:53]: I will add it in another PR, but good idea. :-) Maybe I will discover more!

**Review by FabienPapet — APPROVED** [review]

**Comment by FabienPapet** [comment]
Tests green can we merge it ?

**Review by florine2623 — APPROVED** [review]
Thanks @Hlavtox & @Progi1984 💪 

QA ✅ in that case :)

**Comment by kpodemski** [comment]
thank you all

**Pull request merged**